### PR TITLE
[usbdpi] DPI changes for additional usbdev testing

### PR DESCRIPTION
**Pull request by alees24**
Substantial further work on the DPI model to exercise the device much more thoroughly, fixing a number of issues in its original implementation, and extending it with a view to supporting control transfers properly and exercising abnormal conditions such as bus errors, disconnects etc.
Coding style migrating incrementally towards standard.
Inclusion of symbol diags in waveforms
OTTF software specifies test configuration to DPI model, so a single chip build can run additional (more to be added).

Please provide comments/feedback on structure/style/readability etc; nothing is expected wrt USB knowledge or reproducing test passes etc.

Supersedes #17228 
Thanks -A

Updated: integration of usbdev_stream_test deferred

### CODE REVIEW

**Review by marnovandermaas — COMMENTED** [review]
Thanks for doing all this work! From a reviewing perspective, it would be easier if this PR was split into multiple commits. For example, commits that only change comments or rename variables should be separated from commits that make functional changes. In general making commits as small as possible (while still staying atomic) would be nice.

**Review by vogelpi — COMMENTED** [review]
Thanks @alees24 for your PR. I did now have a first look but given the amount of code it was on a very high level. Also, I don't feel very confident about doing a good job at reviewing C code. I think someone from the software team might be able to give more useful feedback. As you say, one doesn't necessarily need to have a deep understanding of USB to review this.

The main challenge with this PR is really the size: it's nearly 5k lines of code, 10x of what we actually recommend as max. I understand that it is difficult to make smaller changes because there are so many things on the USB front that need to be improved. And you're certainly doing big steps in that direction. So there is nothing wrong about your output. But we should think about how to break up this and follow-up PRs into smaller chunks. It will be much easier and faster to review those. Ideally, I and hopefully someone else can review one or two 500 line PRs every couple of days. For you it will probably also be easier to address feedback and re-test things when working with smaller PRs. WDYT?

File: sw/device/tests/usbdev_test.c
Comment (alees24) [reply sw/device/tests/usbdev_test.c:52]: Yes, that's used; it allows the DPI model to determine the test that is running. Presently usbdev_test and usbdev_stream_test, plus others in future. It also specifies some test parameters.

File: sw/device/tests/usbdev_stream_test.c
Comment (alees24) [reply sw/device/tests/usbdev_stream_test.c:812]: That's test code and has survived too long; I shall keep it locally, but it shouldn't really be in the commit/repo.

File: sw/device/lib/testing/usb_testutils.c
Comment (alees24) [reply sw/device/lib/testing/usb_testutils.c:59]: Indeed, mostly they're apologies for 'work not yet done because it would make the diff larger'.... far reaching changes to make the DPI signal errors to the test software at run time, for example. I'll create issues for the two or three extensive DPI changes still required/planned; that will encompass most of the 'TODO' breadcrumbs I've left

**Comment by alees24** [comment]
> And you're certainly doing big steps in that direction. So there is nothing wrong about your output. But we should think about how to break up this and follow-up PRs into smaller chunks. It will be much easier and faster to review those. Ideally, I and hopefully someone else can review one or two 500 line PRs every couple of days. For you it will probably also be easier to address feedback and re-test things when working with smaller PRs. WDYT?

I could contrive to split the PR into small bites, but I have already tried to separate the new functionality into newly-added files:

- usb_stream_test.c - Required for either streaming test to pass
- usbdpi_stream.c/h - Required for streaming on DPI, but obviously not for FPGA
- stream_test.c - Counterpart of the above
- usb_testutils_diags.c/h - Diagnostic functionality/performance monitoring; not required for function

Presumably it doesn't really help with review to strip out new files from the PR?

Corrections to the DPI model and usb_testutils have been required in order to make the streaming test/functionality work at all, of course, and have been discovered over time. I'm hoping that this creates a stable foundation of passing tests; usbdev_test and usbdev_stream_test in t-l sim at present.

Additional tests can be added independently to that.

**Comment by alees24** [comment]
I've pared it back somewhat by removing diags/test functionality and dropping the host-side application, so now what's left is just what's required to achieve a functioning streaming test that can be run in t-l simulation in a regression run.

File: hw/dv/dpi/usbdpi/usb_transfer.c
Comment (alees24) [reply hw/dv/dpi/usbdpi/usb_transfer.c:11]: Counting down from the highest index to the lowest naturally links each descriptor to the one above it such that they are initially linked in order, and it just feels more elegant to use a while loop than a for loop when downcounting. It can, of course, be done using any of the loop statements.

File: hw/dv/dpi/usbdpi/usb_transfer.c
Comment (alees24) [reply hw/dv/dpi/usbdpi/usb_transfer.c:73]: Later code changes include a higher-level function 'transfer_setup' that does exactly that, as another stepping stone towards supporting higher level Control and Data Transfers properly and without so much replication of code.

File: hw/dv/dpi/usbdpi/usb_transfer.c
Comment (alees24) [reply hw/dv/dpi/usbdpi/usb_transfer.c:158]: The test would need to be > with the expression that you suggest, since the 'true' clause is the failure case in which there is insufficient space for the data.

The reason that I express it in the manner sizeof() - num_bytes may appear pedantic; sizeof() is a 'relatively small' constant value, and the 'num_bytes' field is a member of the 'usbdpi_transfer_t' structure, and thus controlled by the transfer_() code.

If we suppose the precondition '0 <= num_bytes <= sizeof()' is met, ie. the structure is in a valid state, the code works for all values of 'n' and the structure is left in a valid, defined state...'n' is a 32-bit 'size_t' quantity in practice, which means that if addition is used ('num_bytes + n') overflow could cause the code to go awry in the event that 'n' is a very large unsigned value.

**Review by vogelpi — COMMENTED** [review]
Thanks for splitting this big PR into chunks, it substantially simplifies reviewing. I think I've identified some more pieces that should be moved into the other PR (see my comments).

I suggest we have a quick offline discussion to see how we can further reduce the size of PRs and instead increase their frequency. Please let me know when you're available @alees24 .

File: hw/dv/dpi/usbdpi/usbdpi.c
Comment (alees24) [reply hw/dv/dpi/usbdpi/usbdpi.c:122]: Done. Thanks

File: hw/dv/dpi/usbdpi/usbdpi.c
Comment (alees24) [reply hw/dv/dpi/usbdpi/usbdpi.c:32]: Yes, but the state names and numbers (in both the SV and the DPI code) need to match, otherwise the waveform views will be misleading.

File: hw/dv/dpi/usbdpi/usbdpi.h
Comment (alees24) [reply hw/dv/dpi/usbdpi/usbdpi.h:123]: I was undecided what to do here. It's not great that there's a different naming convention in the sw/device tree, and that there are two lists of constants, but the deciding factor was that the Bazel build system prevents the DPI model accessing the device software header files; things are compiled in 'sandboxes.'

I could raise this as a more general issue if it's felt that there should be a shared area in which header files may be provided for use by both ends of the 'sw <-> DPI' pipe, for example, but in the interests of expediency I opted to keep the DPI naming internally consistent?

File: sw/device/lib/testing/usb_testutils.h
Comment (alees24) [reply sw/device/lib/testing/usb_testutils.h:97]: Really it should perhaps be implemented, but it predates me and is not yet required anywhere; I just added documentation to the absent/unused function :)

File: sw/device/lib/testing/usb_testutils_diags.h
Comment (alees24) [reply sw/device/lib/testing/usb_testutils_diags.h]: Addressed. I had not fully separated out the diags (one PR, not yet created) and streaming (another, extant PR) when trying to simplify things.
I definitely want to ensure that the diags code is included at some point, however, since it is excruciating working at the level of t-l sim. The build and simulation times are so long, and visibility into the activity and timing of the software is extremely poor without the diags I have found it necessary to create.
Alas the timings of the sw and the DPI model are still closely interlinked, just less so than earlier.

Comment (alees24) [reply sw/device/lib/testing/usb_testutils_diags.h]: Providing no 'test descriptor' (NULL desc pointer) means that the usb_testutils layer will not respond to the command from the DPI model. That's fine for the hello_usbdev example application at present, since it only runs on the FPGA, but the top-level simulation tests must provide a response or the DPI model will halt; this is more robust than having it make assumptions in the event of invalid or absent response.

Comment (vogelpi) [reply sw/device/lib/testing/usb_testutils_diags.h]: This makes sense, thanks for the clarification @alees24 .

File: hw/dv/dpi/usbdpi/usbdpi.h
Comment (vogelpi) [reply hw/dv/dpi/usbdpi/usbdpi.h:123]: We've briefly discussed this offline and agreed that an issue should be created for this. This seems to be a common use case and mostly likely there is a solution available for this somewhere. Opening an issue helps to start discussing and raise awareness without blocking this PR.

**Review by vogelpi — APPROVED** [review]
Thanks @alees24 for updating and the PR. I think we should move forward with this. There will be follow-up PRs to further improve the USB DPI model and how it's used.

Please do not forget to open the issues as discussed and link them to this PR. We don't need to address these things right now but we also don't want to forget about it.

I've now re-triggered the previously failing CI runs (timeout) and will merge the PR once CI passes.

**Pull request merged**
